File: sprint03/KMS/First week/BOJ_2667.py
n = int(input())
graph = []
visited = [[0]*n for _ in range(n)]
for i in range(n):
    graph.append(list(map(int, input())))



name =2
def dfs(x,y):
    global cnt
    global n
    global name
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    if graph[x][y] == 1:
        cnt = cnt +1
        visited[x][y] = 1
    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        # 범위를 넘었을 때
        if(nx < 0 or ny < 0 or nx >= n or ny >= n):
            continue
        if(graph[nx][ny] != 1):
            continue
        else:
            graph[nx][ny] = name
            if(visited[nx][ny]== 0):
                visited[nx][ny] = 1
                cnt = cnt +1
                dfs(nx,ny)
                
                
result = []
for i in range(n):
    for j in range(n):
        if visited[i][j] == 0 and graph[i][j] == 1:
            name = name+1
            cnt = 0
            dfs(i,j)
            result.append(cnt)
            
        
print(name-2)   

result.sort()
for i in result:
    print(i)

File: sprint03/JMS/BOJ_2667.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

int n, cnt = 0, housecnt;
int arr[25][25];
int diri[4] = { -1,1,0,0 }, dirj[4] = { 0,0,-1,1 };
bool visited[25][25] = { false };
vector<int> housecntV;

typedef struct index {
	int i, j;
}Index;

void Input() {
	cin >> n;

	for (int i = 0; i < n; i++) {
		string temp;
		
		cin >> temp;
		for (int j = 0; j < temp.size(); j++)
			arr[i][j] = temp[j] - '0';
	}
}

void DFS(Index index) {
	visited[index.i][index.j] = true;

	for (int i = 0; i < 4; i++) {
		Index sum = { index.i + diri[i], index.j + dirj[i] };

		if (sum.i < 0 || sum.j < 0 || sum.i >= n || sum.j >= n || visited[sum.i][sum.j] || !arr[sum.i][sum.j])
			continue;

		housecnt++;
		DFS(sum);
	}
}

int main() {
	Input();

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			if (!visited[i][j] && arr[i][j]) {
				cnt++;
				housecnt = 1;
				DFS({ i,j });
				housecntV.push_back(housecnt);
			}
		}
	}

	sort(housecntV.begin(), housecntV.end());

	cout << cnt;
	for (auto p : housecntV)
		cout << endl << p;
}
